camp/camp-brooklyn/src/test/java/org/apache/brooklyn/camp/brooklyn/ConfigInheritanceYamlTest.java

```diff
@@ -300,14 +338,212 @@ public void testOverridesParentConfig() throws Exception {
+                    "  brooklyn.children:",
+                    "  - type: entity-with-keys");
+            
+            Entity app = createStartWaitAndLogApplication(new StringReader(yaml));
+            TestEntity entity = (TestEntity) Iterables.getOnlyElement(app.getChildren());
+    
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-merged")), ImmutableMap.of("myDefaultKey", "myDefaultVal"));
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-always")), ImmutableMap.of("myDefaultKey", "myDefaultVal"));
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-never")), ImmutableMap.of("myDefaultKey", "myDefaultVal"));
+        }
+
+        // Test override defaults in parent entity
+        {
+            String yaml = Joiner.on("\n").join(
+                    "services:",
+                    "- type: org.apache.brooklyn.entity.stock.BasicApplication",
+                    "  brooklyn.config:",
+                    "    map.type-merged:",
+                    "      mykey: myval",
+                    "    map.type-always:",
+                    "      mykey: myval",
+                    "    map.type-never:",
+                    "      mykey: myval",
+                    "  brooklyn.children:",
+                    "  - type: entity-with-keys");
+            
+            Entity app = createStartWaitAndLogApplication(new StringReader(yaml));
+            TestEntity entity = (TestEntity) Iterables.getOnlyElement(app.getChildren());
+    
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-merged")), ImmutableMap.of("mykey", "myval"));
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-always")), ImmutableMap.of("mykey", "myval"));
+            assertEquals(entity.config().get(entity.getEntityType().getConfigKey("map.type-never")), ImmutableMap.of("myDefaultKey", "myDefaultVal"));
+        }
+
+        // Test merging/overriding of explicit config
+        {
+            String yaml = Joiner.on("\n").join(
+                    "services:",
+                    "- type: org.apache.brooklyn.entity.stock.BasicApplication",
+                    "  brooklyn.config:",
+                    "    map.type-merged:",
+                    "      mykey: myval",
+                    "    map.type-always:",
+                    "      mykey: myval",
+                    "    map.type-never:",
+                    "      mykey: myval",
+                    "  brooklyn.children:",
+                    "  - type: entity-with-keys",
+                    "    brooklyn.config:",
+                    "      map.type-merged:",
+                    "        mykey2: myval2",
+                    "      map.type-always:",
+                    "        mykey2: myval2",
+                    "      map.type-never:",
+                    "        mykey2: myval2");
```

Comment:
That's a really good point! I've confirmed that behaviour in a test. In your example, we'd get ENTITY_ID and CLUSTER_NAME, but not APP_NAME.

That feels very hard to fix. We'd have to change the way `EntityConfigMap.inheritedConfig` is populated and stored. Currently the entity is given a copy of its parent's config (rather than looking it up on the parent each time).

In fact, I'm not sure that declaring `brooklyn.parameters` at each level would work around it! The `BasicConfigurationSupport.refreshInheritedConfig` calls `entity.getAllConfig()`. That just does `Map result; result.putAll(inheritedConfig); result.putAll(ownConfig);` so it ignores the config key's settings for inheritance.

I'm tempted to leave that as a documented limitation, with the failing test marked as WIP+Broken. Does that sound sensible to you @neykov ?